data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/client/CloudWatchLogsService.java

```diff
@@ -0,0 +1,221 @@
+     * @param logs - Collection of Record events which hold log data.
+     */
+    public void output(final Collection<Record<Event>> logs) {
+        reentrantLock.lock();
+
+        try {
+            if (!stopWatchOn) {
+                startStopWatch();
+            }
+
+            for (Record<Event> singleLog: logs) {
+                String logJsonString = singleLog.getData().toJsonString();
+                int logLength = logJsonString.length();
+
+                if (thresholdCheck.isGreaterThanMaxEventSize(logLength + LOG_EVENT_OVERHEAD_SIZE)) {
+                    LOG.warn("Event blocked due to Max Size restriction! {Event Size: " + (logLength + LOG_EVENT_OVERHEAD_SIZE) + " bytes}");
+                    continue;
+                }
+
+                int bufferSizeWithOverHead = (buffer.getBufferSize() + (buffer.getEventCount() * LOG_EVENT_OVERHEAD_SIZE));
+                if ((thresholdCheck.isGreaterThanThresholdReached(getStopWatchTime(),  bufferSizeWithOverHead + logLength + LOG_EVENT_OVERHEAD_SIZE, buffer.getEventCount() + 1) && (buffer.getEventCount() > 0))) {
+                    pushLogs();
+                }
+
+                if (singleLog.getData().getEventHandle() != null) {
+                    bufferedEventHandles.add(singleLog.getData().getEventHandle());
+                }
+                buffer.writeEvent(logJsonString.getBytes());
+            }
+
+            runExitCheck();
+
+        } catch (InterruptedException e) {
+            LOG.error("Caught InterruptedException while attempting to publish logs!");
+            reentrantLock.unlock();
+        }
+    }
+
+    private void pushLogs() throws InterruptedException {
+        LOG.info("Attempting to push logs! {Batch size: " + buffer.getEventCount() + "}");
+        stopAndResetStopWatch();
+        startStopWatch();
+
+        ArrayList<InputLogEvent> logEventList = new ArrayList<>();
+        failedPost = true;
+
+        for (byte[] data: buffer.getBufferedData()) {
+            InputLogEvent tempLogEvent = InputLogEvent.builder()
+                    .message(new String(data))
+                    .timestamp(System.currentTimeMillis())
+                    .build();
+            logEventList.add(tempLogEvent);
+        }
+
+        while (failedPost && (failCounter < retryCount)) {
+            try {
+                PutLogEventsRequest putLogEventsRequest = PutLogEventsRequest.builder()
+                        .logEvents(logEventList)
+                        .logGroupName(logGroup)
+                        .logStreamName(logStream)
+                        .build();
+
+                cloudWatchLogsClient.putLogEvents(putLogEventsRequest);
+
+                requestSuccessCount.increment();
+                failedPost = false;
+
+                //TODO: When a log is rejected by the service, we cannot send it, can probably push to a DLQ here.
+
+            } catch (AwsServiceException | SdkClientException e) {
```

Comment:
These exceptions are all throwable by the PutLogEvents API call, but from the documentation I believe only a few of them are actually retryable  by their nature. In our case CloudWatchLogsException is a superclass of most of these exception types so I will refactor to have that one instead of AwsServiceException as AwsServiceException is generalized across services. I believe we don't have to specifically catch all of those unless we wanted to include unique retransmission handling procedures for each distinct call. It might be better to explore this in a future update to the sink as we are mainly focused on retrying when we get an exception we can recover from.